### ISSUE TITLE
supporting maven publishing

### ISSUE THREAD

**Issue by johnynek** [bazelbuild/rules_scala#921]
I've been thinking of a design to support maven publishing. I will sketch it out below.

```
jvm_module_import(
  name = "baz_import",
  deps = [":baz_jar"],
)

scala_library(
  name = "foo",
  srcs = ["Foo.scala"]
)

scala_library(
  name = "bar",
  srcs = ["Bar.scala"],
  deps = [":foo"]
)

jvm_module(
  group = "com.example",
  artifact = "foobar",
  version = "1.0",
  # all these are JavaInfo values and are merged into the current artifact.
  includes = [":foo", ":bar"],
  # the deps are all other jvm modules which have maven coordinates.
  deps = [":baz_import"],
)
```

1. introduce a new provider, such as MavenInfo provider that includes the maven coordinates. We make sure that `scala_import/java_import/etc...` includes these coordinates.
2.add a new rule, such as `scala_module` that takes a list of rules and creates a new MavenInfo. The output here will be: pom.xml, a single jar with all the code in the module and other files required by publishing and a script that when you call `bazel run` for this target it will do the publishing. This can be done via coursier publishing or some other code: https://github.com/coursier/coursier/issues/1489
3. the scala_module when `build` is called does a check that all the transitive dependency jars are either included in the current module, or in a module we depend on. This means we can produce the pom.xml and it will be precise.
4. have an aspect that scans all modules in the repo to ensure that none of them produce jars with duplicate classnames in them, perhaps using classgraph: https://github.com/classgraph/classgraph

related issue: https://github.com/bazelbuild/bazel/issues/1372

Of course, I don't think anything I mentioned above is scala specific. So, you could do this generically as `jvm_module` probably. To do that, you probably need `jvm_module_import` rule to add the MavenInfo provider to JavaInfo values.

**Comment by ittaiz** [comment]
Sounds good.
Few thoughts:
1. Maybe this should be in rules_jvm_external? It’s the other way around then what they currently have but I think it fits in there (cc @jin )
2. We tried to build this super generic service back at the time to help migration and hit roadblocks on translating bazel to maven (neverlink for example) and on solving bazel target depending on maven which depends on maven_exported_from_bazel. Both of these might be irrelevant here but thought I’d share.
3.  From a similar docker experience I think there’s a lot of value in making sure the script can be run outside of Bazel or it’s easy to download the outputs so it can run outside. This requirement will cause issues if people will try to run the script on the CI machine of the build (because it will add a lot of time to the build and not allow to use build without the bytes). Again this might be out of scope for beginning but thought I’d share.

**Comment by ittaiz** [comment]
Stumbled on to this related PR https://github.com/bazelbuild/rules_java/pull/26

**Comment by johnynek** [comment]
I glanced at it.

1. I think it is weird that importing is in rules_java_external but exporting is considered something that should be in rules_java
2. I didn’t see how this does any checking that the published artifact is correct as I proposed above. I think this will lead to many broken poms/jars being published.

**Comment by ittaiz** [comment]
1. I agree (also mentioned in my original comment).
2. Probably you’re right, maybe add your 2c?

**Comment by ittaiz** [comment]
Btw, only concern about 1 is that I’d rather not use rules jvm external for importing but maybe use it for exporting. I’d hate for them to tie it together

**Comment by johnynek** [comment]
I added a comment.

**Comment by amit-traiana** [comment]
Have a look at what [Graknlabs](https://github.com/graknlabs/bazel-distribution) did. We already using it to create a POM file and Publish the artifact for our Java libraries. I'm actually working on a fork that will let it work with `scala-rules` (small change really).

**Comment by ittaiz** [comment]
@amit-traiana the edit didn't send a notification sorry. You can look for `JavaInfo` provider.
@natansil do you want to maybe contribute a readme section on how you were able to publish scala code using rules_scala and graknlabs rules?

**Comment by amitz** [comment]
@ittaiz No worries :-) Found a way around it. I made the change to support Scala, and PR was merged, so Graknlabs solution now support publishing of Scala JARs if that's useful. 
Thanks!

**Comment by ittaiz** [comment]
@natansil used it for publishing our newest OSS'd library from bazel so indeed useful. Probably we have you to thank for that so thank you!
Maybe you want to contribute a readme section about maven publishing?

**Comment by amitz** [comment]
Happy someone else use this code :-)
Contribution to `rules_scala` README? I guess there is not much to say except mentioning Graknlabs project as an option as they have a pretty decent documentation. Unless that's not the documentation you're talking about?

**Comment by ittaiz** [comment]
That sounds valuable. I think that it will allow someone coming to the project to easily see that they can publish to maven while using rules_scala.

**Comment by amitz** [comment]
Sure. I can add a short section to the README for the people who want to publish `rules_scala` artifacts outside of Bazel. Just note it might take a bit of time. Finding it hard to find some free-time with this whole working-from-home-with-kids thingi :)

**Comment by ittaiz** [comment]
Sure thing. Completely understand :)

**Comment by natansil** [comment]
> @natansil do you want to maybe contribute a readme section on how you were able to publish scala code using rules_scala and graknlabs rules?

Sure. I'll add a section soon.

**Comment by ittaiz** [comment]
fixed by documenting existing solution (#1044)
